### PR TITLE
Atomics cleanup

### PR DESCRIPTION
**Pull request by davidchisnall**
Also fixes the build with GCC.

### CODE REVIEW

**Comment by jsonn** [comment]
I'm not sure I consider this a step forward. It certainly makes a lot of things more obscure to me

**Comment by davidchisnall** [comment]
I'm not sure you are reading the same code that this is replacing that I am.  It had grown incredibly baroque with the different macros everywhere and was almost impossible to read.  I had to spend several minutes puzzling over individual lines to figure out what it was doing.  It had interleaved code that was only used in one representation with code that was used everywhere.  It was using atomics incorrectly in several places but those were difficult to spot because they were hidden in layers of macros (and made harder to audit by the fact that variables that must be accessed atomically were not qualified in any way and so could be accidentally used with non-atomic operations).

The new code has separate classes for the two representation.  Each is independently simpler than the interleaved combination and can be read separately.  All uses of atomics are wrapped in a C++11 `std::atomic`-like object and so behave in exactly the same way as any other C++ codebase from the last decade and will always conservatively fall back to sequentially consistent atomics if not explicitly weakened.

File: src/exception.cc [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /* 
  * Copyright 2010-2011 PathScale, Inc. All rights reserved.
+ * Copyright 2021 David Chisnall, Inc. All rights reserved.
```
Comment (MaskRay) [inline src/exception.cc:3]: , Inc? 🙄

Comment (davidchisnall) [reply src/exception.cc:3]:
```suggestion
 * Copyright 2021 David Chisnall. All rights reserved.
```

Ooops, copy and paste fail.

**Pull request merged**